plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/HostedMachineProviderImpl.java

```diff
@@ -167,10 +187,36 @@ protected void pullImage(CheServiceImpl service,
             if (workDir != null) {
                 FileCleaner.addFile(workDir);
             }
+            // When new image is being built it pulls base image. This operation is performed by docker build command.
+            // So, after build it is needed to cleanup base image if it is a snapshot.
+            if (service.getImage().contains(MACHINE_SNAPSHOT_PREFIX)) {
+                submitCleanSnapshotImageTask(service.getImage());
+            }
         }
     }
 
     /**
+     * Sometimes swarm cannot delete image after its pull during a few seconds.
+     * To workaround that problem and avoid redundant delay on workspace start
+     * we must clean up snapshot image in separate thread after some delay.
+     *
+     * @param image
+     *         image to clean, e.g. 172.11.12.13:5000/machine_snapshot_abcdef1234567890:latest
+     */
+    private void submitCleanSnapshotImageTask(String image) {
```

Comment:
because our schedule annotations designed for periodic tasks, unlike this which require to run cleanup task only once after some delay.